### PR TITLE
[RLlib] `MultiAgentEpisode` for Multi-Agent Reinforcement Learning with the new `EnvRunner` API.

### PR DESCRIPTION
**Pull request by simonsays1980**
The `MultiAgentEpisode` should be used as a container to store all episode information of `MultiAgentEnv`s. 

Specifically it has to track which agent stepped at which environmen step.

## Why are these changes needed?

With the change from `RolloutWorker` to `EnvRunner` we switch from `EpisodeV2` to `SingleAgentEpisode` and `MultiAgentEpisode`. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:41]: Can we add a docstring here explaining all the args?

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:133]: Add `Args:` section and `Returns:`.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:175]: Add `Args:` section and `Returns:`.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:217]: Add `Args:` section and `Returns:`.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:262]: Add proper docstring.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:300]: docstring

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:347]: docstring

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:453]: Add `Args:` section and `Returns:`.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:510]: Add `Args:` section and `Returns:`.

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:12]: remove this TODO

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:65]: Can we add `infos` to this test as well?

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:88]: Can we add the `infos` as well to this test?

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:103]: add reset `infos` to this call.

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:98]: Dumb question: I though one would have to do `env.reset(seed=...)` now in gymnasium.
Does it matter, which one we do?

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:116]: for correctness:
`if is_terminated or is_truncated:`

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:136]: Nice!

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:169]: Nice!

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:166]: Can we add another two lines here that test, whether if we (hackish'ly) add stuff the the successor's lists (manually), whether it influences the original episode (it shouldn't)?

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [inline rllib/env/tests/test_single_agent_episode.py:219]: Can we quickly assert here that `episode_2` is untouched after this concatenation and also check some of the indices (like assert that `episode_1.observations[105]` is the equal (but not the same object) as `episode_2.observations[5]`)?

File: rllib/utils/replay_buffers/episode_replay_buffer.py
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:363]: Nice!

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:9]: Nit: Can we move this class to the bottom of the file? I know we still do this a lot in RLlib, but we should always have the main class come first in its file and any helper classes moved to the bottom.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:10]: Can you explain more what this will be used for?

Plus add docstring to the `find_indices` method.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [reply rllib/env/multi_agent_episode.py:9]: Also, we should make this a private class (simply rename to `_IndexMapping`).

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:20]: Remove this TODO: We are just hours to merging the PPO EnvRunner PR.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:98]: I like this API thus far. :)

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:178]: Any chance we can unify this entire logic here in `get_observations|rewards|actions` into one or more helper utilities?

**Review by sven1977 — COMMENTED** [review]
Looks all very good already! :)
Just a few missing docstrings and other nits.

Note: I'd be happy to merge this (once comments are addressed) and do the multi-agent episode tests in a follow up PR to not explode this one too much.

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:65]: As soon as the ENvRunner PR is merged and can merge the master into this PR and add `infos` and `extra_model_outputs`.

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:88]: As soon as the ENvRunner PR is merged and can merge the master into this PR and add `infos` and `extra_model_outputs`.

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:103]: As soon as the ENvRunner PR is merged and can merge the master into this PR and add `infos` and `extra_model_outputs`.

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:98]: I try this out. This might be the better option. Thanks for the hint @sven1977

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:166]: I put this in. Thanks for this idea @sven1977

File: rllib/env/tests/test_single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/tests/test_single_agent_episode.py:219]: Great idea @sven1977 ! I will implement this!

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:9]: No problem! Consider it done.

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:20]: Yeah this will resolve a lot of TODOs then. Looking forward!

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:98]: Yeah me too. ALso using SAE in MAE - a lot is simply passing through

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:178]: I am positive. Give me some time to think about it

**Comment by simonsays1980** [comment]
> Looks all very good already! :) Just a few missing docstrings and other nits.
> 
> Note: I'd be happy to merge this (once comments are addressed) and do the multi-agent episode tests in a follow up PR to not explode this one too much.

@sven1977 Yes I like this approach. I will focus on the comments and make the MAE test in another one

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py]: Refactored into a single helper function `_getattr_by_index`

File: rllib/algorithms/ppo/ppo.py
Comment (sven1977) [inline rllib/algorithms/ppo/ppo.py:454]: Nice! Thanks for adding the typehints.

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:15]: Nice!

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:20]: nit: `Optional[int]`

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:23]: nit: `Optional[List[np.ndarray]]`

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:24]: nit: optional ...

File: rllib/env/single_agent_episode.py
Comment (sven1977) [inline rllib/env/single_agent_episode.py:519]: nit:

```
    Returns:
        An integer, defining the length of an episode.

    Raises:
        AssertionError: If ....
```

File: rllib/env/tests/test_single_agent_episode.py
Comment (sven1977) [reply rllib/env/tests/test_single_agent_episode.py:326]: I think you are right. It's impossible for episode_2's data objects to be used in the now elongated episode_1.
Ok, fine with me. Let's remove these comments here and that's it :)

File: rllib/algorithms/dreamerv3/utils/env_runner.py
Comment (simonsays1980) [reply rllib/algorithms/dreamerv3/utils/env_runner.py:28]: @sven1977 Great catch!

Comment (simonsays1980) [reply rllib/algorithms/dreamerv3/utils/env_runner.py:28]: Resolved

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:122]: Actually, this is not correct (I think this came from DreamerV3 debugging :( ):
"the images include the corresponding rewards."

File: rllib/algorithms/ppo/ppo.py
Comment (simonsays1980) [reply rllib/algorithms/ppo/ppo.py:454]: I thought this is better readable for users, to see from whereon we switch to `SampleBatch`.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:133]: Let's remove this comment. We shouldn't assume thing as to how this API is used.

Let's add instead:
- docstring
- the rules for when it's ok to concatenate:
- - IDs are identical
- - timesteps match (`t` of `self` matches `t_started` of `episode_chunk`)

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:196]: Nice!!

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:220]: Super nice!! :)

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:243]: :)

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:257]: nit: add initial_state and initial_render_image.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:296]: docstring missing

File: rllib/env/single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/single_agent_episode.py:23]: ANother great one :) Resolved in next commit.

File: rllib/env/single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/single_agent_episode.py:24]: Resolved.

File: rllib/env/single_agent_episode.py
Comment (simonsays1980) [reply rllib/env/single_agent_episode.py:519]: Alright. Have changed this.

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:122]: Alright I will remove this from all `render_images` docstrings.

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:340]: Let's explain this here a little (what we discussed on slack):

```
For MultiAgentEpisode: self.is_terminated should be True, if all agents are terminated and self.is_truncated should be True if all agents are truncated.
If only one or more agents (but not all!) are terminated or truncated, the MultiAgentEpisode.is_terminated/is_truncated should both be False. The information about single agents' terminated/truncated states can always be retrieved from the SingleAgentEpisodes inside the MultiAgent one.
If all agents are either terminated or truncated (but in a mixed fashion: some agents terminated, some truncated): This is currently undefined and could potentially be a problem (if a user really implemented such a multi-agent env that behaves this way). My guess is that we should probably then set both is_terminated AND is_truncated in the MultiAgentEpisode to True. Question here: Does this have practical relevance?
```

File: rllib/env/multi_agent_episode.py
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:29]: I think what you said on slack is correct:


Simon:
```
But was with the case The environment is not terminated and some agents are terminated earlier on, e.g. timestep 10 agent 1 temrinated (because it fell down a cliff) the other agents are still alive at timestep 100. Do we need here probably in the _init_  of the MultiAgentEpisode`  is_terminated and is_terminated as a list of `MultiAgentDict`s like [{...}, ... , {"agent_1": True, "agent_2": False, ...}, ....] such that we know which agent is still alive and where agent 1 died?

Sven:
```
Great point! Maybe we should NOT allow users to construct a MA Episode via observations and actions and ... But rather by providing a bunch of agentIDs mapping to SingleAgentEpisodes?
```

Comment (sven1977) [reply rllib/env/multi_agent_episode.py:29]: Just to keep this in mind: Thus far, initializing a new episode (single- or multi-agent) with already existing data is an edge case, that we have - to the best of my knowledge - never used in RLlib. So it's not a super urgent decision we need to make, just to keep this in the back of our heads.

Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:29]: See my comment to this on slack. There might be a hurdle with MA episodes.

File: rllib/env/multi_agent_episode.py
Comment (simonsays1980) [reply rllib/env/multi_agent_episode.py:340]: I think it has for postprocessing: If truncated, the `GAE` uses for example bootstrapping, while if `terminated` the value of $o_T$ is set to `0.0`.

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks for this PR, @simonsays1980 ! Let's wait for tests to pass, then we can merge this.

**Pull request merged**
